app/src/main/java/com/asfoundation/wallet/ui/backup/BackupCreationPresenter.kt

```diff
@@ -1,42 +1,183 @@
+            cachedFileName = it
+            activityView.openSystemFileDirectory(it)
+          } else {
+            handleDialogSaveClickBelowAndroidQ(it)
+          }
+        }
+        .subscribe())
   }
 
-  private fun handleBackupClick() {
+  private fun handleDialogSaveClickBelowAndroidQ(fileName: String) {
+    disposables.add(fileInteract.createAndSaveFile(cachedKeystore, downloadsPath, fileName)
+        .subscribeOn(networkScheduler)
+        .observeOn(viewScheduler)
+        .doOnComplete {
+          view.closeDialog()
+          activityView.showSuccessScreen()
+        }
+        .doOnError { showError(it) }
+        .subscribe({}, { showError(it) }))
+  }
+
+  private fun handleDialogCancelClick() {
+    disposables.add(view.getDialogCancelClick()
+        .doOnNext { view.closeDialog() }
+        .subscribe())
+  }
+
+  private fun createBackUpFile() {
+    disposables.add(exportWalletInteract.export(walletAddress, password)
+        .doOnSuccess { cachedKeystore = it }
+        .flatMapCompletable { fileInteract.createTmpFile(walletAddress, it, temporaryPath) }
+        .subscribeOn(networkScheduler)
+        .observeOn(viewScheduler)
+        .doOnComplete { view.enableSaveButton() }
+        .doOnError { showError(it) }
+        .subscribe({}, { showError(it) }))
+  }
+
+  private fun handlePositiveButtonClick() {
     disposables.add(view.getPositiveButtonClick()
+        .observeOn(viewScheduler)
         .doOnNext {
           if (fileShared) {
             activityView.closeScreen()
           } else {
-            view.shareFile("")
-            fileShared = true
+            val file = fileInteract.getCachedFile()
+            if (file == null) showError("Error retrieving file")
+            else {
+              context?.let { view.shareFile(fileInteract.getUriFromFile(context, file)) }
+                  ?: showError("Null context")
+              fileShared = true
+            }
           }
         }
         .subscribe())
   }
 
-  private fun handleCloseBtnClick() {
-    disposables.add(view.getNegativeButtonClick()
+  private fun handleSaveAgainClick() {
+    disposables.add(view.getSaveAgainClick()
+        .doOnNext { activityView.startWalletBackup() }
+        .subscribe())
+  }
+
+  private fun handlePermissionGiven() {
+    disposables.add(activityView.onPermissionGiven()
         .doOnNext {
+          view.showSaveOnDeviceDialog(fileInteract.getDefaultBackupFileFullName(walletAddress),
+              downloadsPath?.path)
+        }
+        .subscribe({}, { it.printStackTrace() }))
+  }
+
+  private fun createAndSaveFile(documentFile: DocumentFile,
+                                fileName: String): Completable {
+    return fileInteract.createAndSaveFile(cachedKeystore, documentFile, fileName)
+        .observeOn(viewScheduler)
+        .doOnComplete {
+          view.closeDialog()
           activityView.showSuccessScreen()
         }
-        .subscribe())
+        .doOnError { showError(it) }
```

Comment:
is this doOnError needed here? we are catching the error already in handleSystemFileIntentResult()